## Упражнения по библиотеке Numpy

In [2]:
import numpy as np

**1.** Дан случайный массив, поменять знак у элементов, значения которых между 3 и 8

In [88]:
a = np.random.randint(1, 10, size = 10) 
print(a)

mask = (a > 3) & (a < 8)

a[mask] = -a[mask]
        
print("result:\n", a)

[6 7 8 4 3 5 9 6 9 6]
result:
 [-6 -7  8 -4  3 -5  9 -6  9 -6]


**2.** Заменить максимальный элемент случайного массива на 0

In [87]:
a = np.random.randint(1, 10, size = 10) 

print(a)

a[np.argmax(a)] = 0

print("result:\n", a)

[2 5 4 7 8 1 5 3 1 3]
result:
 [2 5 4 7 0 1 5 3 1 3]


**3.** Построить прямое произведение массивов (все комбинации с каждым элементом). На вход подается двумерный массив

In [96]:
a = np.array(np.random.randint(1, 10, size = (2, 2)))
print(a)

n_rows, n_cols = a.shape
# все комбинации индексов
indices = np.indices((n_rows,) * n_cols)
indices = indices.reshape(n_cols, -1).T

result = np.zeros((n_rows ** n_cols, n_cols), dtype=int)
for i, idx_combo in enumerate(indices):
    result[i] = a[idx_combo, range(n_cols)]

print("result\n", result)

[[9 4]
 [8 3]]
result
 [[9 4]
 [9 3]
 [8 4]
 [8 3]]


**4.** Даны 2 массива A (8x3) и B (2x2). Найти строки в A, которые содержат элементы из каждой строки в B, независимо от порядка элементов в B

In [103]:
A = np.array([
    [9, 1, 4],
    [1, 3, 4],
    [7, 9, 6],
    [8, 4, 3],
    [4, 6, 6],
    [6, 3, 9],
    [2, 7, 2],
    [5, 6, 9]
])
B = np.array([
    [3, 4],
    [1, 9]
])

masks = []
for row_b in B:
    # проверка для каждого элемента B есть ли он в A
    element_masks = [np.any(A == element, axis=1) for element in row_b]
    
    # строка A должна содержать хотя бы один элемент из текущей строки B
    masks.append(np.any(element_masks, axis=0))

final_mask = np.all(masks, axis=0)
result = A[final_mask]

print('matching rows:')
print(result)


matching rows:
[[9 1 4]
 [1 3 4]
 [6 3 9]]


**5.** Дана 10x3 матрица, найти строки из неравных значений (например строка [2,2,3] остается, строка [3,3,3] удаляется)

In [20]:
arr = np.array([
    [1, 1, 1],
    [2, 3, 4],
    [4, 6, 8],
    [4, 4, 4],
    [5, 8, 2],
    [5, 5, 5],
    [9, 10, 11],
    [9, 9, 9],
    [2, 2, 2],
    [0, 0, 0]
])

mask = ~np.all(arr == arr[:, [0]], axis = 1)
    
print('result\n', arr[mask])    


result
 [[ 2  3  4]
 [ 4  6  8]
 [ 5  8  2]
 [ 9 10 11]]


**6.** Дан двумерный массив. Удалить те строки, которые повторяются

In [109]:
arr = np.array([
    [1, 1, 1],
    [2, 3, 4],
    [4, 6, 8],
    [4, 4, 4],
    [1, 1, 1],
    [4, 6, 8],
    [0, 0, 0]
])

arr_unique, count  = np.unique(arr, axis=0, return_counts=True)

result = arr_unique[count == 1]

print('result\n', result)



result
 [[0 0 0]
 [2 3 4]
 [4 4 4]]


______
______

Для каждой из следующих задач (1-5) нужно привести 2 реализации – одна без использования numpy (cчитайте, что там, где на входе или выходе должны быть numpy array, будут просто списки), а вторая полностью векторизованная с использованием numpy (без использования питоновских циклов/map/list comprehension).


__Замечание 1.__ Можно считать, что все указанные объекты непустые (к примеру, в __задаче 1__ на диагонали матрицы есть ненулевые элементы).

__Замечание 2.__ Для большинства задач решение занимает не больше 1-2 строк.

___

* __Задача 1__: Подсчитать произведение ненулевых элементов на диагонали прямоугольной матрицы.  
 Например, для X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]) ответ 3.

In [85]:
# without numpy
X = [[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]

print(X)

rows = len(X)
cols = len(X[0])

size_diag = min(rows, cols)

result = 1

# проходим по диагональным элементам
for i in range(size_diag):
    if X[i][i] != 0:
        result *= X[i][i]
    
print("result: ", result)

[[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]
result:  3


In [ ]:
# numpy

X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]])

diagonal = X.diagonal()

result = np.prod(diagonal[diagonal != 0])

print(result)


* __Задача 2__: Даны два вектора x и y. Проверить, задают ли они одно и то же мультимножество.  
  Например, для x = np.array([1, 2, 2, 4]), y = np.array([4, 2, 1, 2]) ответ True.

In [104]:
# without numpy

x = [1, 2, 2, 4]
y = [4, 2, 1, 2]

x.sort()
y.sort()

if x == y:
    print(True)
else:
    print(False)


True


In [102]:
#numpy

x = np.array([1, 2, 2, 4])
y = np.array([4, 2, 1, 2])

if np.array_equal(np.sort(x), np.sort(y)):
    print(True)
else:
    print(False)


True


* __Задача 3__: Найти максимальный элемент в векторе x среди элементов, перед которыми стоит ноль. 
 Например, для x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0]) ответ 5.

In [126]:
# without numpy

x = [6, 2, 0, 3, 0, 0, 5, 7, 0]

after_zero = []

# ищем эл-ты, которые стоят после нуля
for i in range(1, len(x)):
    if x[i-1] == 0:
        after_zero.append(x[i])
        
print(max(after_zero))


5


In [127]:
# numpy

x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0])

mask = (x[:-1] == 0)

print(np.max(x[1:][mask]))

5


* __Задача 4__: Реализовать кодирование длин серий (Run-length encoding). Для некоторого вектора x необходимо вернуть кортеж из двух векторов одинаковой длины. Первый содержит числа, а второй - сколько раз их нужно повторить.  
 Например, для x = np.array([2, 2, 2, 3, 3, 3, 5]) ответ (np.array([2, 3, 5]), np.array([3, 3, 1])).

In [106]:
## without numpy

x = [2, 2, 2, 3, 3, 3, 5, 2, 1]

nums, count = [], []

if x:
    current_num = x[0]   # начинаем с первого элемента
    current_count = 1

    for i in range(1, len(x)):
        if x[i] == current_num:
            current_count += 1  # увеличиваем счетчик для текущего числа
        else:
            # сохраняем предыдущую серию и начинаем новую
            nums.append(current_num)
            count.append(current_count)
            current_num = x[i]
            current_count = 1
    
    # добавляем последнюю серию
    nums.append(current_num)
    count.append(current_count)
        
print(nums)
print(count)


[2, 3, 5, 2, 1]
[3, 3, 1, 1, 1]


In [107]:
# numpy

x = np.array([2, 2, 2, 3, 3, 3, 5, 6, 6, 7, 2])

# Находим индексы, где элементы меняются
change_indices = np.where(x[1:] != x[:-1])[0] + 1

# Границы серий: начало каждой серии
starts = np.concatenate(([0], change_indices))

# Значения - первые элементы каждой серии
nums = x[starts]

# Длины серий
count = np.diff(np.concatenate((starts, [len(x)])))

print(nums)
print(count)

[2 3 5 6 7 2]
[3 3 1 2 1 1]


In [ ]:
# Находим индексы, где элементы меняются
change_indices = np.where(x[1:] != x[:-1])[0] + 1

# Границы серий: начало каждой серии
starts = np.concatenate(([0], change_indices))

# Значения - первые элементы каждой серии
nums = x[starts]

# Длины серий
counts = np.diff(np.concatenate((starts, [len(x)])))

In [ ]:
# Находим индексы, где элементы меняются
change_indices = np.where(x[1:] != x[:-1])[0] + 1

# Границы серий: начало каждой серии
starts = np.concatenate(([0], change_indices))

# Значения - первые элементы каждой серии
nums = x[starts]

# Длины серий
counts = np.diff(np.concatenate((starts, [len(x)])))

* __Задача 5__: Даны две выборки объектов - X и Y. Вычислить матрицу евклидовых расстояний между объектами. Сравните с функцией scipy.spatial.distance.cdist по скорости работы.

In [81]:
import scipy.spatial.distance
import time

X = np.random.rand(100, 10).tolist()
Y = np.random.rand(80, 10).tolist()

print(f"Размеры: X {len(X)}x{len(X[0])}, Y {len(Y)}x{len(Y[0])}")


Размеры: X 100x10, Y 80x10


In [82]:
# without numpy

start_time = time.time()
n_X = len(X)
n_Y = len(Y)
result_no_numpy = [[0.0] * n_Y for _ in range(n_X)]
    
for i in range(n_X):
    for j in range(n_Y):
        # Вычисляем евклидово расстояние между X[i] и Y[j]
        squared_distance = 0
        for k in range(len(X[i])):
            squared_distance += (X[i][k] - Y[j][k]) ** 2
        result_no_numpy[i][j] = (squared_distance) ** 0.5
time_no_numpy = time.time() - start_time

print(f"Первые 2x2 элемента без numpy:")
for i in range(2):
    print([f"{result_no_numpy[i][j]:.4f}" for j in range(2)])


Первые 2x2 элемента без numpy:
['1.0797', '1.5855']
['1.2585', '1.1916']


In [83]:
X = np.array(X)
Y = np.array(Y)

start_time = time.time()

# расширяем размерности массивов для попарного сравнения всех элементов
X_expanded = X[:, np.newaxis, :]  # # каждая строка X сравнивается с каждой строкой Y
Y_expanded = Y[np.newaxis, :, :]  # создаем "копии" Y для каждой строки X
    
# вычисляем матрицу расстояний
squared_diff = np.sum((X_expanded - Y_expanded) ** 2, axis=2)
result_numpy = np.sqrt(squared_diff)
time_numpy = time.time() - start_time

print(f"\nПервые 2x2 элемента с numpy:")
print(result_numpy[:2, :2])

start_time = time.time()
result_scipy = scipy.spatial.distance.cdist(X, Y, metric='euclidean')
time_scipy = time.time() - start_time

print(f"\nПервые 2x2 элемента scipy:")
print(result_scipy[:2, :2])



Первые 2x2 элемента с numpy:
[[1.07974531 1.58546752]
 [1.25850581 1.19163877]]

Первые 2x2 элемента scipy:
[[1.07974531 1.58546752]
 [1.25850581 1.19163877]]


In [84]:
print('Время работы')
print(f"Без numpy: {time_no_numpy:.6f} сек")
print(f"С numpy:   {time_numpy:.6f} сек")
print(f"Scipy:     {time_scipy:.6f} сек")

Время работы
Без numpy: 0.090468 сек
С numpy:   0.002440 сек
Scipy:     0.001020 сек


_______
________

* #### __Задача 6__: CrunchieMunchies __*__

Вы работаете в отделе маркетинга пищевой компании MyCrunch, которая разрабатывает новый вид вкусных, полезных злаков под названием **CrunchieMunchies**.

Вы хотите продемонстрировать потребителям, насколько полезны ваши хлопья по сравнению с другими ведущими брендами, поэтому вы собрали данные о питании нескольких разных конкурентов.

Ваша задача - использовать вычисления Numpy для анализа этих данных и доказать, что ваши **СrunchieMunchies** - самый здоровый выбор для потребителей.


In [23]:
import numpy as np

1. Просмотрите файл cereal.csv. Этот файл содержит количества калорий для различных марок хлопьев. Загрузите данные из файла и сохраните их как calorie_stats.

In [24]:
calorie_stats = np.loadtxt("./data/cereal.csv", delimiter=",")
calorie_stats

array([ 70., 120.,  70.,  50., 110., 110., 110., 130.,  90.,  90., 120.,
       110., 120., 110., 110., 110., 100., 110., 110., 110., 100., 110.,
       100., 100., 110., 110., 100., 120., 120., 110., 100., 110., 100.,
       110., 120., 120., 110., 110., 110., 140., 110., 100., 110., 100.,
       150., 150., 160., 100., 120., 140.,  90., 130., 120., 100.,  50.,
        50., 100., 100., 120., 100.,  90., 110., 110.,  80.,  90.,  90.,
       110., 110.,  90., 110., 140., 100., 110., 110., 100., 100., 110.])

2. В одной порции CrunchieMunchies содержится 60 калорий. Насколько выше среднее количество калорий у ваших конкурентов?

Сохраните ответ в переменной average_calories и распечатайте переменную в терминале

In [52]:
averange_calories = calorie_stats.mean(axis=0) - 60
round(averange_calories, 2)

46.88

3. Корректно ли среднее количество калорий отражает распределение набора данных? Давайте отсортируем данные и посмотрим.

Отсортируйте данные и сохраните результат в переменной calorie_stats_sorted. Распечатайте отсортированную информацию

In [27]:
calorie_stats_sorted = np.sort(calorie_stats, axis=0)
calorie_stats_sorted

array([ 50.,  50.,  50.,  70.,  70.,  80.,  90.,  90.,  90.,  90.,  90.,
        90.,  90., 100., 100., 100., 100., 100., 100., 100., 100., 100.,
       100., 100., 100., 100., 100., 100., 100., 100., 110., 110., 110.,
       110., 110., 110., 110., 110., 110., 110., 110., 110., 110., 110.,
       110., 110., 110., 110., 110., 110., 110., 110., 110., 110., 110.,
       110., 110., 110., 110., 120., 120., 120., 120., 120., 120., 120.,
       120., 120., 120., 130., 130., 140., 140., 140., 150., 150., 160.])

4. Похоже, что большинство значений выше среднего. Давайте посмотрим, является ли медиана наиболее корректным показателем набора данных.

Вычислите медиану набора данных и сохраните свой ответ в median_calories. Выведите медиану, чтобы вы могли видеть, как она сравнивается со средним значением.

In [31]:
median_calories = np.median(calorie_stats, axis=0)
median_calories

110.0

5. В то время как медиана показывает, что по крайней мере половина наших значений составляет более 100 калорий, было бы более впечатляюще показать, что значительная часть конкурентов имеет более высокое количество калорий, чем CrunchieMunchies.

Рассчитайте различные процентили и распечатайте их, пока не найдете наименьший процентиль, превышающий 60 калорий. Сохраните это значение в переменной nth_percentile.

In [48]:
i = 0
nth_percentile_value = 0

while nth_percentile_value < 60 and i < 100:
    i += 1
    nth_percentile_value = np.percentile(calorie_stats, i)
    print(f"{i}%: {nth_percentile_value:.1f}")
    
nth_percentile = nth_percentile_value

nth_percentile
    

1%: 50.0
2%: 50.0
3%: 55.6
4%: 70.0


70.0

6. Хотя процентиль показывает нам, что у большинства конкурентов количество калорий намного выше, это неудобная концепция для использования в маркетинговых материалах.

Вместо этого давайте подсчитаем процент хлопьев, в которых содержится более 60 калорий на порцию. Сохраните свой ответ в переменной more_calories и распечатайте его

In [49]:
more_calories = 100 - i
more_calories


96

7. Это действительно высокий процент. Это будет очень полезно, когда мы будем продвигать CrunchieMunchies. Но один вопрос заключается в том, насколько велики различия в наборе данных? Можем ли мы сделать обобщение, что в большинстве злаков содержится около 100 калорий или разброс еще больше?

Рассчитайте величину отклонения, найдя стандартное отклонение, Сохраните свой ответ в calorie_std и распечатайте на терминале. Как мы можем включить эту ценность в наш анализ?

In [51]:
calorie_std = np.std(calorie_stats)
round(calorie_std,2)

19.36

8. Напишите короткий абзац, в котором кратко изложите свои выводы и то, как, по вашему мнению, эти данные могут быть использованы в интересах Mycrunch при маркетинге CrunchieMunchies.

**Вывод**

1. Калорийность печенья CrunchieMunchies существенно меньше чем у его конкурентов (60.0 калорий против медианного значения 110.0)
2. Калорийность CrunchieMunchies на 44.88% ниже, чем средний продукт на рынке
3. Большинство конкурентов калорийней CrunchieMunchies: 96% продуктов на рынке содержат более 60 калорий на порцию. Это значит, что CrunchieMunchies менее калорийны, чем практически все аналогичные продукты
4. Стандартное отклонение 19.36 калорий показывает, что рынок достаточно однороден по калорийности, и наше преимущество является статистически значимым

Для эффективной маркетинговой кампании продвижения CrunchieMunchies необходимо сделать акцент на том, что CrunchieMunchies содержат на 45% меньше калорий, 
  чем средний продукт на рынке. Подчеркнуть, что 96% продуктов на рынке содержат более, чем 60 калорий на порцию. Использовать яркие слоганы и визуализации, чтобы показать положение CrunchieMunchies на фоне конкурентов

